cmd/lncli/chainrpc_active.go

```diff
@@ -0,0 +1,200 @@
+//go:build chainrpc
+// +build chainrpc
+
+package main
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"strconv"
+
+	"github.com/btcsuite/btcd/wire"
+	"github.com/lightningnetwork/lnd/lnrpc/chainrpc"
+	"github.com/urfave/cli"
+)
+
+// chainCommands will return the set of commands to enable for chainrpc builds.
+func chainCommands() []cli.Command {
+	return []cli.Command{
+		{
+			Name:     "chain",
+			Category: "On-chain",
+			Usage:    "Interact with the bitcoin blockchain.",
+			Subcommands: []cli.Command{
+				getBlockCommand,
+				getBestBlockCommand,
+				getBlockHashCommand,
+			},
+		},
+	}
+}
+
+func getChainClient(ctx *cli.Context) (chainrpc.ChainKitClient, func()) {
+	conn := getClientConn(ctx, false)
+
+	cleanUp := func() {
+		conn.Close()
+	}
+
+	return chainrpc.NewChainKitClient(conn), cleanUp
+}
+
+var getBlockCommand = cli.Command{
+	Name:        "getblock",
+	Category:    "On-chain",
+	Usage:       "Get block by block hash.",
+	Description: "Returns a block given the corresponding block hash.",
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "hash",
+			Usage: "the target block hash",
+		},
+		cli.BoolFlag{
+			Name:  "verbose",
+			Usage: "print entire block as JSON",
+		},
+	},
+	Action: actionDecorator(getBlock),
+}
+
+func getBlock(ctx *cli.Context) error {
+	ctxc := getContext()
+
+	var (
+		args            = ctx.Args()
+		blockHashString string
+	)
+
+	verbose := false
+	if ctx.IsSet("verbose") {
+		verbose = true
+	}
+
+	switch {
+	case ctx.IsSet("hash"):
+		blockHashString = ctx.String("hash")
+
+	case args.Present():
+		blockHashString = args.First()
+
+	default:
+		return fmt.Errorf("hash argument missing")
+	}
+
+	blockHash, err := hex.DecodeString(blockHashString)
+	if err != nil {
+		return err
+	}
+
+	client, cleanUp := getChainClient(ctx)
+	defer cleanUp()
+
+	req := &chainrpc.GetBlockRequest{BlockHash: blockHash}
+	resp, err := client.GetBlock(ctxc, req)
+	if err != nil {
+		return err
+	}
+
+	// Convert raw block bytes into wire.MsgBlock.
+	msgBlock := &wire.MsgBlock{}
+	blockReader := bytes.NewReader(resp.RawBlock)
+	err = msgBlock.Deserialize(blockReader)
+	if err != nil {
+		return err
+	}
+
+	if verbose {
+		printJSON(msgBlock)
+	} else {
+		printJSON(msgBlock.Header)
+	}
+
+	return nil
+}
+
+var getBestBlockCommand = cli.Command{
+	Name:     "getbestblock",
+	Category: "On-chain",
+	Usage:    "Get best block.",
+	Description: "Returns the latest block hash and height from the valid " +
+		"most-work chain.",
+	Action: actionDecorator(getBestBlock),
+}
+
+func getBestBlock(ctx *cli.Context) error {
+	ctxc := getContext()
+
+	client, cleanUp := getChainClient(ctx)
+	defer cleanUp()
+
+	resp, err := client.GetBestBlock(ctxc, &chainrpc.GetBestBlockRequest{})
+	if err != nil {
+		return err
+	}
+
+	printRespJSON(resp)
```

Comment:
This is very annoying about the whole raw bytes vs. string representation of chain hashes (TXIDs and block hashes): This output will show the block hash in the reverse byte order. So I can't copy/paste it into a block explorer. We need to manually reverse the bytes before printing the JSON. See `exportChanBackup` for an example where we create an anonymous struct with JSON tags and then use the `String()` method of the `chainhash.Hash` to show the hash in the correct order.

Reply:
Add `MarshalJSON()` to `chainhash.Hash` https://github.com/btcsuite/btcd/pull/1932 ?

Reply:
That sounds good. But the problem is that on this level you don't have a `chainhash.Hash` but a gRPC bytes slice.

Reply:
Added code to extract `chainhash.Hash` from gRPC in latest commits.